protocol/swap/manager.go

```diff
@@ -1,168 +1,156 @@
-func (i *Info) ProvidedAmount() float64 {
-	return i.providedAmount
-}
+var _ Manager = &manager{}
 
-// ReceivedAmount returns the amount of coin received for this swap, in standard units.
-func (i *Info) ReceivedAmount() float64 {
-	return i.receivedAmount
-}
-
-// ExchangeRate returns the exchange rate for this swap, represented by a ratio of XMR/ETH.
-func (i *Info) ExchangeRate() types.ExchangeRate {
-	return i.exchangeRate
-}
-
-// EthAsset returns the Ethereum asset for this swap, either an ERC-20 address or the zero address
-// for regular ETH
-func (i *Info) EthAsset() types.EthAsset {
-	return i.ethAsset
-}
-
-// Status returns the swap's status.
-func (i *Info) Status() Status {
-	return i.status
-}
-
-// StatusCh returns the swap's status update channel.
-func (i *Info) StatusCh() <-chan types.Status {
-	return i.statusCh
-}
-
-// SetStatus ...
-func (i *Info) SetStatus(s Status) {
-	i.status = s
-}
+var errNoSwapWithID = errors.New("unable to find swap with given ID")
 
 // Manager tracks current and past swaps.
 type Manager interface {
 	AddSwap(info *Info) error
-	GetPastIDs() []types.Hash
-	GetPastSwap(types.Hash) *Info
-	GetOngoingSwap(types.Hash) *Info
-	CompleteOngoingSwap(types.Hash)
+	GetPastIDs() ([]types.Hash, error)
+	GetPastSwap(types.Hash) (*Info, error)
+	GetOngoingSwap(types.Hash) (*Info, error)
+	CompleteOngoingSwap(types.Hash) error
 }
 
 type manager struct {
+	db Database
 	sync.RWMutex
 	ongoing map[types.Hash]*Info
 	past    map[types.Hash]*Info
 }
 
-// NewManager ...
-func NewManager() Manager {
+// NewManager returns a new Manager that uses the given database.
+// It loads all ongoing swaps into memory on construction.
+// Completed swaps are not loaded into memory.
+func NewManager(db Database) (*manager, error) {
+	ongoing := make(map[types.Hash]*Info)
+
+	stored, err := db.GetAllSwaps()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, s := range stored {
+		if !s.Status.IsOngoing() {
+			continue
+		}
+
+		ongoing[s.ID] = s
+	}
+
 	return &manager{
-		ongoing: make(map[types.Hash]*Info),
+		db:      db,
+		ongoing: ongoing,
 		past:    make(map[types.Hash]*Info),
-	}
+	}, nil
 }
 
 // AddSwap adds the given swap *Info to the Manager.
 func (m *manager) AddSwap(info *Info) error {
 	m.Lock()
 	defer m.Unlock()
 
-	switch info.status.IsOngoing() {
+	switch info.Status.IsOngoing() {
 	case true:
-		m.ongoing[info.id] = info
+		m.ongoing[info.ID] = info
 	default:
-		m.past[info.id] = info
+		m.past[info.ID] = info
 	}
 
-	return nil
+	return m.db.PutSwap(info)
 }
 
 // GetPastIDs returns all past swap IDs.
-func (m *manager) GetPastIDs() []types.Hash {
+func (m *manager) GetPastIDs() ([]types.Hash, error) {
 	m.RLock()
 	defer m.RUnlock()
-	ids := make([]types.Hash, len(m.past))
-	i := 0
+	ids := make(map[types.Hash]struct{})
 	for id := range m.past {
-		ids[i] = id
+		ids[id] = struct{}{}
+	}
+
+	// TODO: do we want to cache all past swaps since we're already fetching them?
```

Comment:
We might need a max results parameter and a way to sort the past swaps from most recent to least recent. Even if they all fit in memory, the user may want to limit the number retrieved.

Reply:
I'll make an issue for this! might need to update `swap.Info` to have a `completed time.Time` or something